src/odemis/util/registration.py

```diff
@@ -0,0 +1,464 @@
+    # `(x₁, x₂) * (n, m)ᵀ = n * x₁ + m * x₂`.
+    matrix = _canonical_matrix_form(numpy.transpose(dxy))
+
+    # The estimated translated is the centroid of the input coordinates.
+    translation = numpy.mean(xy, axis=0)
+
+    return AffineTransform(matrix, translation)
+
+
+def estimate_grid_orientation(
+    xy: numpy.ndarray, shape: Tuple[int, int], transform_type: Type[T]
+) -> T:
+    """
+    Estimate the orientation of a square grid of points.
+
+    Parameters
+    ----------
+    xy : ndarray of shape (n, 2)
+        Array with the determined coordinates of the grid points.
+    shape : tuple of two ints
+        The shape of the grid given as a tuple `(height, width)`. Current
+        implementation only supports grids of points with `height == width`.
+    transform_type : GeometricTransform
+        The transform class to use for estimating the orientation.
+
+    Returns
+    -------
+    tform : GeometricTransform
+        The orientation of the pattern.
+
+    """
+    if shape[0] != shape[1]:
+        raise NotImplementedError(
+            "Only grids with `shape[0] == shape[1]` are supported."
+        )
+    tform = _initial_estimate_grid_orientation(xy)
+    grid = unit_gridpoints(shape, mode="xy")
+    correspondences = bijective_matching(tform.apply(grid), xy)
+    a, b = zip(*correspondences)
+    return transform_type.from_pointset(grid[a, ...], xy[b, ...])
+
+
+def estimate_grid_orientation_from_img(
+    image: numpy.ndarray,
+    shape: Tuple[int, int],
+    transform_type: Type[T],
+    sigma: float,
+    threshold_abs: Optional[float] = None,
+    threshold_rel: Optional[float] = None,
+    num_spots: Optional[int] = None,
+) -> T:
+    """
+    Image based estimation of the orientation of a square grid of points.
+
+    Parameters
+    ----------
+    image : ndarray
+        The input image.
+    shape : tuple of two ints
+        The shape of the grid given as a tuple `(height, width)`.
+    transform_type : GeometricTransform
+        The transform class to use for estimating the orientation.
+    sigma : float
+        Expected size of the spots. Assuming the spots are Gaussian shaped,
+        this is the standard deviation.
+    threshold_abs : float, optional
+        Minimum intensity of peaks. By default, the absolute threshold is
+        the minimum intensity of the image.
+    threshold_rel : float, optional
+        If provided, apply a threshold on the minimum intensity of peaks,
+        calculated as `max(image) * threshold_rel`.
+    num_spots : int, optional
+        Maximum number of spots. When the number of spots exceeds `num_spots`,
+        return `num_spots` peaks based on highest spot intensity. Will use
+        `num_spots = shape[0] * shape[1]` as default when set to `None`. Set
+        `num_spots = 0` to not impose a maximum. Note that this behavior is
+        different from odemis.util.spot.find_spot_position().
+
```

Comment:
can you add an explanation about what it returns? And is the returned transform a square grid, or can it contain shear etc?